libraries/freertos_plus/standard/tls/src/iot_tls.c

```diff
@@ -122,9 +122,11 @@ static void prvFreeContext( TLSContext_t * pxCtx )
         mbedtls_ssl_free( &pxCtx->xMbedSslCtx );
         mbedtls_ssl_config_free( &pxCtx->xMbedSslConfig );
 
-        /* Cleanup PKCS#11. */
-        if( ( NULL != pxCtx->pxP11FunctionList ) &&
-            ( NULL != pxCtx->pxP11FunctionList->C_CloseSession ) )
+        /* Cleanup PKCS11 only if the handshake succeeded. */
+        if( ( pdTRUE == pxCtx->xTLSHandshakeSuccessful ) &&
```

Comment:
prvFreeContext can be called as a result of a handshake error. In that case, there will be a P11 context, but HandshakeSuccessful will be false.